### ISSUE TITLE
Evaluation suite - Segmentation metrics

### ISSUE THREAD

**Issue by csudre** [Project-MONAI/MONAI#889]
**Context**
By default more than one metric should be used for any type of evaluation - For instance, Dice score should never be used on its own

**Describe the solution you'd like**
- All confusion matrix metrics [https://en.wikipedia.org/wiki/Sensitivity_and_specificity](https://en.wikipedia.org/wiki/Sensitivity_and_specificity)
- Hausdorff distance with argument for percentile
- Average distance
- Surface dice

**Additional context**
Additional information is available [https://github.com/Project-MONAI/MONAI/wiki/Evaluation-metrics-task-force](https://github.com/Project-MONAI/MONAI/wiki/Evaluation-metrics-task-force)
@AReinke @jenspetersen @danieltudosiu

**Comment by wyli** [comment]
I think we could start working on this and report back any issues (especially if they are related to #887 #888) cc @Nic-Ma @yiheng-wang-nv

**Comment by Nic-Ma** [comment]
Sounds good! I will go through all the plan with @yiheng-wang-nv today.

Thanks.

**Comment by Nic-Ma** [comment]
Hi @wyli ,

I totally agree to add more metrics as soon as possible, we are really short of metrics.
And I think most of the metrics are common or standard algorithms, should we re-implement them all in MONAI or leverage some 3rd packages?
When more metrics included, let's try to enhance metrics module with a detailed report and other cool features from working group.

Thanks.

**Comment by csudre** [comment]
Hi @wyli, @Nic-Ma !

As mentioned on the wiki page, there are a few implementations that can potentially be used:
[https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.directed_hausdorff.html](url)
[https://scikit-learn.org/stable/modules/model_evaluation.html](url)
[https://medium.com/pytorch/pytorch-lightning-metrics-35cb5ab31857](url)
[https://github.com/NifTK/NiftyNet/tree/dev/niftynet/evaluation](url)
[https://github.com/cocodataset/cocoapi/blob/master/PythonAPI/pycocotools/cocoeval.py](url)

Thanks!

**Comment by wyli** [comment]
I think we could leverage the relatively mature packages such as scipy/scikit-learn as a starting point. We can always take more iterations to refactor them if there are concerns from the working group.

**Comment by Nic-Ma** [comment]
Hi @wyli ,

I agree with scipy/scikit-learn for common metrics, and @yiheng-wang-nv will start from there.
And for medical specific metrics, let's discuss one by one whether to use 3rd package or port to MONAI.

Thanks.

**Comment by Nic-Ma** [comment]
Mark: @yiheng-wang-nv is working these metrics proposals, will submit several PRs.
Thanks.

**Comment by Nic-Ma** [comment]
Hi @yiheng-wang-nv , please assign this ticket to yourself.

Thanks.

**Comment by yiheng-wang-nv** [comment]
> Hi @yiheng-wang-nv , please assign this ticket to yourself.
> 
> Thanks.

Hi @csudre @wyli could you please help to assign this ticket to me? Here I may not have the authority to self-assign

**Comment by rijobro** [comment]
Hi, has there been any more progress on this one (beyond #916)?

If not, I was thinking about giving it a go.

**Comment by Nic-Ma** [comment]
Hi @rijobro ,

Thanks for your interest, of course, glad to see your contribution here, which metrics do you want to develop?
@yiheng-wang-nv is working some other task right now, will also continue to develop some soon.

Thanks.

**Comment by rijobro** [comment]
OK, I guess I'll start with any mentioned above and go from there. I think progress may be slow as I'm still getting familiar with the codebase.

@yiheng-wang-nv - If you let me know when you start work on this again, I can let you know where I'm up to.

**Comment by yiheng-wang-nv** [comment]
Hi @rijobro , I just finished other tasks and will continue to this task now. Should I start with Surface dice?

**Comment by rijobro** [comment]
Yeah feel free to continue with any of the others. A few of the metrics might need surface comparisons, in which case it might be best to move [this](https://github.com/Project-MONAI/MONAI/blob/master/monai/metrics/hausdorff_distance.py#L72-L78) functionality to a common location (e.g., `monai/utils`). Up to you.

**Comment by wyli** [comment]
these low-level metric computations are now in place.

**Comment by JanSellner** [comment]
Hey, I just wanted to ask what is the status about the [normalized surface dice](https://arxiv.org/abs/1809.04430) metric which is mentioned in the issue description? I couldn't find it in the documentation. Is it possible that this issue was closed too early? (maybe @wyli or @yiheng-wang-nv can comment on this). Thank you and keep up the great work! :-)

**Comment by wyli** [comment]
I believe we missed that one -- reopening this for surface Dice

**Comment by NabJa** [comment]
Hi, any updates on the [surface dice](https://arxiv.org/abs/1809.04430) implementation? I'd like to give it a go if no one is working on it.

**Comment by wyli** [comment]
> Hi, any updates on the [surface dice](https://arxiv.org/abs/1809.04430) implementation? I'd like to give it a go if no one is working on it.

great, please go ahead!
